src/wp-includes/taxonomy.php

```diff
@@ -3333,13 +3381,201 @@ function wp_defer_term_counting( $defer = null ) {
 		$_defer = $defer;
 		// Flush any deferred counts.
 		if ( ! $defer ) {
+			wp_modify_term_count_by( null, null, null, true );
 			wp_update_term_count( null, null, true );
 		}
 	}
 
 	return $_defer;
 }
 
+/**
+ * Prevents add/removing a term from modifying a term count.
+ *
+ * This is used by functions calling wp_transition_post_status() to indicate the
+ * term count will be handled during the post's transition.
+ *
+ * @private
+ * @since 5.6.0
+ *
+ * @param bool $new_setting The new setting for preventing term counts.
+ * @return bool Whether term count prevention is enabled or disabled.
+ */
+function _wp_prevent_term_counting( $new_setting = null ) {
+	static $prevent = false;
+
+	if ( is_bool( $new_setting ) ) {
+		$prevent = $new_setting;
+	}
+
+	return $prevent;
+}
```

Comment:
This is what broke it in the 5.6 cycle (see Trac ticket 51630), it was a little too eager about preventing term counting and caused the count to be incorrect in some circumstances.